### PR TITLE
Add support for AMD GPUs

### PR DESCRIPTION
**Pull request by jakurzak**
## Overview
This pull request adds support for AMD Instinct GPUs.

### `CMakeLists.txt`
- The newly added section of the CMake script introduces logic to detect the presence of the hipcc compiler, which is part of AMD's ROCm software platform. If the hipcc compiler is detected, CMake will configure the build system to compile the HIP source files using hipcc.

### `Makefile`
- Add `HIPCC=hipcc` and `HIPCCFLAGS = -O3` variables, similarly to the existing `NVCC=nvcc` and `HIPCCFLAGS = -O3` variables.
- Add `.PHONY: qsim-hip`, `.PHONY: hip-tests`, and `.PHONY: run-hip-tests` sections, similarly to the existing `.PHONY: qsim-cuda`, `.PHONY: cuda-tests`, and `.PHONY: run-cuda-tests` sections.

### `apps/Makefile`
- Generate the list of `HIP_TARGETS` wih the `.hip.x` suffix from all files with the `.cuda.cu` suffix, similarly to how the list of `CUDA_TARGETS` is generated.
- Add the `.PHONY: qsim-hip` and the `%hip.x: %cuda.cu` section, similarly to the existing `.PHONY: qsim-cuda` and `%cuda.x: %cuda.cu` sections.

### `apps/make.sh`
- The added section compiles CUDA source files with the hipcc compiler instead of nvcc when hipcc is available.

### `docs/_book.yaml`
- Add "AMD GPU support" article to the "Other tutorials" section.

### `docs/tutorials/amd_gpu.md`
- Add basic documentation for the AMD GPU support.

### `lib/cuda2hip.h`
- Define a series of macros that map CUDA function calls, types, constants, and utilities to their corresponding HIP counterparts.
- Provide a template function `__shfl_down_sync()`, as a wrapper around HIP's `__shfl_down()` function, since the former is not currently available in HIP.

### `lib/fuser_mqubit.h`
- Remove unused variable to eliminate Clang warning.

### `lib/simulator_cuda_kernels.h`
- Add a section for including HIP headers if the HIP compiler is detected.

### `lib/statespace_cuda.h`
- Add a section for including HIP headers if the HIP compiler is detected.
- Add a missing error check for `cudaMemset()`.

### `lib/statespace_cuda_kernels.h`
- Add a section for including HIP headers if the HIP compiler is detected.

### `lib/util_cuda.h`
- Add a section for including HIP headers if the HIP compiler is detected.

### `lib/vectorspace_cuda.h`
- Add a section for including HIP headers if the HIP compiler is detected.
- Add missing error checks for `cudaFree()`, `cudaMemcpy()`, and `cudaDeviceSynchronize()`.

### `pybind_interface/Makefile`
- Add `QSIMLIB_HIP`, similar to the existing `QSIMLIB_CUDA`.
- Rename `pybind-gpu` and `decide-gpu` to `pybind-cuda` and `decide-cuda`.
- Add `pybind-hip` and `decide-hip`.
- Add removal of HIP files in the `clean` section.

### `pybind_interface/decide/CMakeLists.txt`
- To add HIP support to the CMake file, checks were introduced to detect whether hipcc is available. If hipcc is found, the project is configured to use HIP by setting `HIP` as one of the project languages. The CMake file then adjusts the build process to use HIP-specific commands (`hip_add_library`) and properties for compiling the relevant source files. It also includes the necessary directories and sets the module properties for creating a Python extension module with HIP support, allowing the same build system to work across both CUDA and HIP platforms depending on the available compiler.

### `pybind_interface/decide/decide.cpp`
- Add `HIP = 2`  to  `enum GPUCapabilities`.
- Set `GPUCapabilities gpu = HIP` if HIP compiler is detected.

### `pybind_interface/hip/CMakeLists.txt`
- Set up HIP compilation for a Python extension module written in HIP and C++. Configure compiler optimizations and OpenMP settings. Integrate Pybind11 for C++/Python interfacing, and locate necessary Python libraries.

### `pybind_interface/hip/pybind_main_hip.cpp`
- Integrate with Pybind11 for creating Python bindings by creating a factory pattern for object creation and setting up the necessary environment.

### `pybind_interface/hip/pybind_main_hip.h`
- Introduce Python bindings for a module called `qsim_hip` using Pybind11.

### `pybind_interface/pybind_main.cpp`
- Remove unused lambda captures: `circuit`, `ncircuit`, and `state_space`.

### `qsimcirq/__init__.py`
- Add detection of HIP. If `qsim_decide.detect_gpu()` returns 2, import the HIP-compatible module `qsimcirq.qsim_hip`.

### `setup.py`
- Add `import shutil`.
- If hipcc is found, append arguments to the `cmake_args` list to specify hipcc as the C and C++ compiler.
- Specify that the `qsim_hip` directory contains a CMake project to be built as a Python extension with support for HIP.

### `tests/Makefile`
- Add rules for building HIP tests, similar to the rules for building CUDA tests.

### `tests/make.sh`
- Add rules for building HIP tests, similar to the rules for building CUDA tests.

## Building

```
make -j qsim      # to build CPU qsim
make -j qsim-hip  # to build HIP qsim
make -j pybind    # to build Python bindings
make -j cxx-tests # to build CPU tests
make -j hip-tests # to build HIP tests
pip install .
```

### Simulator

```
make run-cxx-tests # to run CPU tests
make run-hip-tests # to run HIP tests
```

or

```
cd tests
for file in *.x; do ./"$file"; done          # to run all tests
for file in *_hip_test.x; do ./"$file"; done # to run HIP tests only
```

### Python bindings

```
make run-py-tests
```

or

```
cd qsimcirq_tests
python3 -m pytest -v qsimcirq_test.py
```

### CODE REVIEW

**Comment by tanujkhattar** [comment]
@jakurzak Can you please add a PR description ? 

@sergeisakov Can you take a look at this PR?

**Review by sergeisakov — CHANGES_REQUESTED** [review]
Thank you for adding support for AMD GPUs. I can't really test if it works or not. Nevertheless the PR looks good to me. Could you fix a few formatting issues? Could you also add some doc?

**Comment by jakurzak** [comment]
Thank you for taking the time to look at the PR!
I am out of office this week and traveling overseas.
I'll do my best to address the comments ASAP when I'm back in office.

File: lib/statespace_cuda.h [hunk 1]
```diff
@@ -102,7 +107,7 @@ class StateSpaceCUDA :
   }
 
   void SetAllZeros(State& state) const {
-    cudaMemset(state.get(), 0, MinSize(state.num_qubits()) * sizeof(fp_type));
+    ErrorCheck(cudaMemset(state.get(), 0, MinSize(state.num_qubits()) * sizeof(fp_type)));
```
[the remark this answers is not in the document]
Comment (jakurzak) [reply lib/statespace_cuda.h]: Fixed

File: pybind_interface/Makefile [hunk 2]
```diff
@@ -58,12 +68,25 @@ pybind-custatevec:
 decide-custatevec:
 	$(NVCC) decide/decide.cpp -D__CUSTATEVEC__ -o $(QSIMLIB_DECIDE) $(NVCCFLAGS) $(PYBINDFLAGS_CUDA)
 
+.PHONY: pybind-hip
+pybind-hip:
+	$(HIPCC) basic/pybind_main_basic.cpp -o $(QSIMLIB_BASIC) $(CXXFLAGS) $(PYBINDFLAGS_BASIC)
```
[the remark this answers is not in the document]
Comment (jakurzak) [reply pybind_interface/Makefile:73]: Mixing compilers produces a mismatch between Python and C++ data types.
E.g., there's an issue with casting a Python instance of type qsimcirq.qsim_avx2.Circuit to its corresponding C++ type.
`RuntimeError: Unable to cast Python instance of type <class 'qsimcirq.qsim_avx2.Circuit'> to C++ type 'qsim::Circuit<qsim::Gate<float, qsim::Cirq::GateKind> >'`

File: lib/vectorspace_cuda.h [hunk 3]
```diff
@@ -114,29 +119,32 @@ class VectorSpaceCUDA {
       return false;
     }
 
-    cudaMemcpy(dest.get(), src.get(),
-               sizeof(fp_type) * Impl::MinSize(src.num_qubits()),
-               cudaMemcpyDeviceToDevice);
+    ErrorCheck(
+      cudaMemcpy(dest.get(), src.get(),
```
[the remark this answers is not in the document]
Comment (jakurzak) [reply lib/vectorspace_cuda.h]: Fixed

File: lib/vectorspace_cuda.h [hunk 4]
```diff
@@ -114,29 +119,32 @@ class VectorSpaceCUDA {
...
   }
 
   // It is the client's responsibility to make sure that dest has at least
   // Impl::MinSize(src.num_qubits()) elements.
   bool Copy(const Vector& src, fp_type* dest) const {
-    cudaMemcpy(dest, src.get(),
-               sizeof(fp_type) * Impl::MinSize(src.num_qubits()),
-               cudaMemcpyDeviceToHost);
+    ErrorCheck(
+      cudaMemcpy(dest, src.get(),
```
[the remark this answers is not in the document]
Comment (jakurzak) [reply lib/vectorspace_cuda.h]: Fixed

File: lib/vectorspace_cuda.h [hunk 5]
```diff
@@ -114,29 +119,32 @@ class VectorSpaceCUDA {
...
   }
 
   // It is the client's responsibility to make sure that src has at least
   // Impl::MinSize(dest.num_qubits()) elements.
   bool Copy(const fp_type* src, Vector& dest) const {
-    cudaMemcpy(dest.get(), src,
-               sizeof(fp_type) * Impl::MinSize(dest.num_qubits()),
-               cudaMemcpyHostToDevice);
+    ErrorCheck(
+      cudaMemcpy(dest.get(), src,
```
[the remark this answers is not in the document]
Comment (jakurzak) [reply lib/vectorspace_cuda.h]: Fixed

File: lib/vectorspace_cuda.h [hunk 6]
```diff
@@ -145,12 +153,13 @@ class VectorSpaceCUDA {
   // min(size, Impl::MinSize(dest.num_qubits())) elements.
   bool Copy(const fp_type* src, uint64_t size, Vector& dest) const {
     size = std::min(size, Impl::MinSize(dest.num_qubits()));
-    cudaMemcpy(dest.get(), src, sizeof(fp_type) * size, cudaMemcpyHostToDevice);
+    ErrorCheck(
+      cudaMemcpy(dest.get(), src, sizeof(fp_type) * size, cudaMemcpyHostToDevice));
```
[the remark this answers is not in the document]
Comment (jakurzak) [reply lib/vectorspace_cuda.h]: Fixed

**Comment by jakurzak** [comment]
I fixed the style issues and commented on the use of HIPCC for CPU code when compiling Python bindings.
I'll be looking into documentation now.

**Comment by jakurzak** [comment]
I added basic documentation for AMD GPU support.
I added detailed description to the PR.
Please let me know if anything else needs to be done before the PR can be accepted.

File: pybind_interface/Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (sergeisakov) [reply pybind_interface/Makefile:73]: Thanks. It makes sense.

**Comment by sergeisakov** [comment]
> I added basic documentation for AMD GPU support. I added detailed description to the PR. Please let me know if anything else needs to be done before the PR can be accepted.

Thank you. Could you replace Qsim with qsim in `amd_gpu.md`? Could you also add something about how to install the build tools? Is it ROCm?

**Comment by jakurzak** [comment]
> Thank you. Could you replace Qsim with qsim in `amd_gpu.md`? Could you also add something about how to install the build tools? Is it ROCm?

done

**Review by sergeisakov — APPROVED** [review]
LGTM

**Comment by jakurzak** [comment]
Looks like it's getting hung up on the kokoro test.
Any clue why?

**Comment by 95-martin-orion** [comment]
Hmm. Doesn't look like I can re-trigger the kokoro test specifically. Could you push a commit so Github picks up on it?

**Comment by jakurzak** [comment]
No cuQuantum support in the AMD port, BTW, as there's no AMD equivalent.
The HIP port should not affect anything in the CUDA implementation, though.
Pushed a commit to re-trigger.

**Comment by jakurzak** [comment]
@sergeisakov, @95-martin-orion, can you help move this forward?
Thanks in advance!

**Pull request merged**